File: backend/src/main/java/org/jdk/project/controller/VideoController.java
package org.jdk.project.controller;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import lombok.Getter;
import lombok.Setter;
import org.jdk.project.dto.video.VideoStreamDto;
import org.jdk.project.dto.video.VideoStreamsResponse;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/rest/video")
public class VideoController {

  private static final DateTimeFormatter DATETIME_FORMATTER =
      DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

  private final Map<String, VideoStreamDto> activeStreams = new ConcurrentHashMap<>();

  @GetMapping("/active")
  public VideoStreamsResponse activeStreams() {
    return response("活跃视频流已加载");
  }

  @PostMapping("/start")
  public VideoStreamsResponse startStream(@RequestBody VideoStreamRequest request) {
    String robotId = trimToDefault(request.getRobotId(), "ROBOT-001");
    String userId = trimToDefault(request.getUserId(), "admin");
    String now = LocalDateTime.now().format(DATETIME_FORMATTER);
    VideoStreamDto stream =
        VideoStreamDto.builder()
            .userId(userId)
            .mode("mock")
            .startTime(now)
            .frameCount(0)
            .lastFrameAt(now)
            .build();
    activeStreams.put(robotId, stream);
    return response("视频流已启动");
  }

  @PostMapping("/stop")
  public VideoStreamsResponse stopStream(@RequestBody VideoStreamRequest request) {
    String robotId = trimToDefault(request.getRobotId(), "");
    if (robotId.isEmpty()) {
      activeStreams.clear();
      return response("全部视频流已停止");
    }
    if (!robotId.isEmpty()) {
      activeStreams.remove(robotId);
    }
    return response("视频流已停止");
  }

  private VideoStreamsResponse response(String message) {
    return VideoStreamsResponse.builder()
        .data(new LinkedHashMap<>(activeStreams))
        .msg(message)
        .build();
  }

  private String trimToDefault(String value, String defaultValue) {
    if (value == null || value.trim().isEmpty()) {
      return defaultValue;
    }
    return value.trim();
  }

  @Getter
  @Setter
  public static class VideoStreamRequest {
    private String robotId;
    private String userId;
  }
}

File: backend/src/main/kotlin/org/jdk/project/controller/VideoController.kt
package org.jdk.project.controller

import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.concurrent.ConcurrentHashMap
import org.jdk.project.dto.video.VideoStreamDto
import org.jdk.project.dto.video.VideoStreamsResponse
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/rest/video")
class VideoController {
    private val activeStreams: MutableMap<String, VideoStreamDto> = ConcurrentHashMap()

    @GetMapping("/active")
    fun activeStreams(): VideoStreamsResponse = response("活跃视频流已加载")

    @PostMapping("/start")
    fun startStream(@RequestBody request: VideoStreamRequest): VideoStreamsResponse {
        val robotId = trimToDefault(request.robotId, "ROBOT-001")
        val userId = trimToDefault(request.userId, "admin")
        val now = LocalDateTime.now().format(DATETIME_FORMATTER)
        val stream =
            VideoStreamDto(
                userId = userId,
                mode = "mock",
                startTime = now,
                frameCount = 0,
                lastFrameAt = now,
            )
        activeStreams[robotId] = stream
        return response("视频流已启动")
    }

    @PostMapping("/stop")
    fun stopStream(@RequestBody request: VideoStreamRequest): VideoStreamsResponse {
        val robotId = trimToDefault(request.robotId, "")
        if (robotId.isEmpty()) {
            activeStreams.clear()
            return response("全部视频流已停止")
        }
        activeStreams.remove(robotId)
        return response("视频流已停止")
    }

    private fun response(message: String): VideoStreamsResponse =
        VideoStreamsResponse(
            data = linkedMapOf<String, VideoStreamDto>().apply { putAll(activeStreams) },
            msg = message,
        )

    private fun trimToDefault(value: String?, defaultValue: String): String {
        if (value == null || value.trim().isEmpty()) {
            return defaultValue
        }
        return value.trim()
    }

    data class VideoStreamRequest(
        var robotId: String? = null,
        var userId: String? = null,
    )

    private companion object {
        private val DATETIME_FORMATTER: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
    }
}
